oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java

```diff
@@ -273,52 +298,16 @@ protected AccessToken exchange3PICredentialForAccessToken(
     return response.getAccessToken();
   }
 
-  /**
-   * Attempts service account impersonation.
-   *
-   * @param accessToken the access token to be included in the request.
-   * @return the access token returned by the generateAccessToken call.
-   * @throws IOException if the service account impersonation call fails.
-   */
-  protected AccessToken attemptServiceAccountImpersonation(AccessToken accessToken)
-      throws IOException {
-    if (serviceAccountImpersonationUrl == null) {
-      return accessToken;
-    }
-
-    HttpRequest request =
-        transportFactory
-            .create()
-            .createRequestFactory()
-            .buildPostRequest(
-                new GenericUrl(serviceAccountImpersonationUrl),
-                new UrlEncodedContent(new GenericData().set("scope", scopes.toArray())));
-    request.setParser(new JsonObjectParser(OAuth2Utils.JSON_FACTORY));
-    request.setHeaders(
-        new HttpHeaders()
-            .setAuthorization(
-                String.format("%s %s", AuthHttpConstants.BEARER, accessToken.getTokenValue())));
-
-    HttpResponse response;
-    try {
-      response = request.execute();
-    } catch (IOException e) {
-      throw new IOException(
-          String.format("Error getting access token for service account: %s", e.getMessage()), e);
-    }
+  private static String extractTargetPrincipal(String serviceAccountImpersonationUrl) {
+    // Extract the target principle.
+    int startIndex = serviceAccountImpersonationUrl.lastIndexOf('/');
+    int endIndex = serviceAccountImpersonationUrl.indexOf(":generateAccessToken");
 
-    GenericData responseData = response.parseAs(GenericData.class);
-    String token =
-        OAuth2Utils.validateString(responseData, "accessToken", "Expected to find an accessToken");
-
-    DateFormat format = new SimpleDateFormat(RFC3339);
-    String expireTime =
-        OAuth2Utils.validateString(responseData, "expireTime", "Expected to find an expireTime");
-    try {
-      Date date = format.parse(expireTime);
-      return new AccessToken(token, date);
-    } catch (ParseException e) {
-      throw new IOException("Error parsing expireTime: " + e.getMessage());
+    if (startIndex != -1 && endIndex != -1 && startIndex < endIndex) {
+      return serviceAccountImpersonationUrl.substring(startIndex + 1, endIndex);
+    } else {
+      throw new IllegalArgumentException(
+          "Unable to determine target principal from service account impersonation URL.");
```

Comment:
This method is being called lazily (once the credentials are being used). This will be a weird runtime exception thrown that we should catch while building/instantiating the credentials.

Reply:
This is actually called in the constructor (from initializeImpersonatedCredentials)

Reply:
For a confidence check, can we make sure to have a test to ensure this is happening early (at instantiation time).

Reply:
Done!